File: app/core/pagination.py
"""Shared pagination helpers for list APIs."""

from __future__ import annotations

from math import ceil
from typing import Any, Generic, TypeVar

from fastapi import Query
from pydantic import BaseModel, Field, computed_field

DEFAULT_PAGE_SIZE = 25
MAX_PAGE_SIZE = 500
PAGE_SIZE_OPTIONS = (25, 50, 100)

T = TypeVar("T")


class PaginatedResponse(BaseModel, Generic[T]):
    items: list[T]
    total: int
    page: int = Field(ge=1)
    page_size: int = Field(ge=1)
    pages: int = Field(ge=1)

    @classmethod
    def build(
        cls,
        *,
        items: list[T],
        total: int,
        page: int,
        page_size: int,
    ) -> PaginatedResponse[T]:
        pages = max(1, ceil(total / page_size)) if page_size else 1
        return cls(
            items=items,
            total=total,
            page=page,
            page_size=page_size,
            pages=pages,
        )

    @computed_field  # type: ignore[prop-decorator]
    @property
    def total_records(self) -> int:
        return self.total

    @computed_field  # type: ignore[prop-decorator]
    @property
    def total_pages(self) -> int:
        return self.pages

    @computed_field  # type: ignore[prop-decorator]
    @property
    def has_next(self) -> bool:
        return self.page < self.pages

    @computed_field  # type: ignore[prop-decorator]
    @property
    def has_previous(self) -> bool:
        return self.page > 1


class PaginationParams(BaseModel):
    page: int = Field(default=1, ge=1)
    page_size: int = Field(default=DEFAULT_PAGE_SIZE, ge=1, le=MAX_PAGE_SIZE)
    sort: str | None = None
    skip: int = Field(default=0, ge=0)
    limit: int = Field(default=DEFAULT_PAGE_SIZE, ge=1, le=MAX_PAGE_SIZE)

    @classmethod
    def from_query(
        cls,
        *,
        page: int = 1,
        page_size: int | None = None,
        skip: int | None = None,
        limit: int | None = None,
        sort: str | None = None,
    ) -> PaginationParams:
        """Resolve page/page_size with backward-compatible skip/limit."""
        if skip is not None or limit is not None:
            resolved_skip = skip if skip is not None else 0
            resolved_limit = limit if limit is not None else DEFAULT_PAGE_SIZE
            resolved_page = (resolved_skip // resolved_limit) + 1 if resolved_limit else 1
            return cls(
                page=resolved_page,
                page_size=resolved_limit,
                sort=sort,
                skip=resolved_skip,
                limit=resolved_limit,
            )
        resolved_page_size = page_size if page_size is not None else DEFAULT_PAGE_SIZE
        resolved_skip = (page - 1) * resolved_page_size
        return cls(
            page=page,
            page_size=resolved_page_size,
            sort=sort,
            skip=resolved_skip,
            limit=resolved_page_size,
        )


def pagination_query(
    page: int = Query(1, ge=1),
    page_size: int | None = Query(None, ge=1, le=MAX_PAGE_SIZE),
    skip: int | None = Query(None, ge=0),
    limit: int | None = Query(None, ge=1, le=MAX_PAGE_SIZE),
    sort: str | None = Query(None),
) -> PaginationParams:
    return PaginationParams.from_query(
        page=page,
        page_size=page_size,
        skip=skip,
        limit=limit,
        sort=sort,
    )


def apply_sort(stmt: Any, model: type, sort: str | None) -> Any:
    """Apply optional sort field (prefix '-' for descending)."""
    if not sort:
        return stmt
    descending = sort.startswith("-")
    field_name = sort[1:] if descending else sort
    if not hasattr(model, field_name):
        return stmt
    column = getattr(model, field_name)
    return stmt.order_by(column.desc() if descending else column.asc())

File: frontend/src/types/pagination.ts
import type { ListParams } from '../api/client';

export const DEFAULT_PAGE_SIZE = 25;
export const PAGE_SIZE_OPTIONS = [25, 50, 100] as const;
export const PAGE_SIZE_STORAGE_KEY = 'protrack:pagination:page-size';

export interface PaginatedResponse<T> {
  items: T[];
  total: number;
  page: number;
  page_size: number;
  pages: number;
  total_records?: number;
  total_pages?: number;
  has_next?: boolean;
  has_previous?: boolean;
}

export interface PaginationParams extends ListParams {
  page?: number;
  page_size?: number;
  skip?: number;
  limit?: number;
  sort?: string;
}

function asRecord(value: unknown): Record<string, unknown> | null {
  return typeof value === 'object' && value !== null ? (value as Record<string, unknown>) : null;
}

/** Coerce any list API payload into a safe array — never throws. */
export function ensureArray<T>(value: unknown): T[] {
  if (Array.isArray(value)) {
    return value;
  }

  const record = asRecord(value);
  if (!record) {
    return [];
  }

  if (Array.isArray(record.items)) {
    return record.items as T[];
  }
  if (Array.isArray(record.data)) {
    return record.data as T[];
  }
  if (Array.isArray(record.results)) {
    return record.results as T[];
  }

  return [];
}

export function isPaginatedResponse<T>(value: unknown): value is PaginatedResponse<T> {
  const record = asRecord(value);
  if (!record || !Array.isArray(record.items)) {
    return false;
  }

  return (
    typeof record.total === 'number' ||
    typeof record.total_records === 'number'
  );
}

export function unwrapListResponse<T>(value: unknown): T[] {
  return ensureArray<T>(value);
}

export function toSkipLimit(page: number, pageSize: number): { skip: number; limit: number } {
  return {
    skip: (page - 1) * pageSize,
    limit: pageSize,
  };
}

export function clampPage(page: number, total: number, pageSize: number): number {
  const pages = Math.max(1, Math.ceil(total / pageSize));
  return Math.min(Math.max(1, page), pages);
}

export function normalizePaginatedResponse<T>(value: unknown): PaginatedResponse<T> {
  if (isPaginatedResponse<T>(value)) {
    const record = value as PaginatedResponse<T>;
    const total = record.total_records ?? record.total ?? record.items.length;
    const pageSize = record.page_size || DEFAULT_PAGE_SIZE;
    const pages =
      record.total_pages ??
      record.pages ??
      Math.max(1, Math.ceil(total / pageSize));
    const page = record.page || 1;

    return {
      items: record.items,
      total,
      page,
      page_size: pageSize,
      pages,
      total_records: total,
      total_pages: pages,
      has_next: record.has_next ?? page < pages,
      has_previous: record.has_previous ?? page > 1,
    };
  }

  const items = ensureArray<T>(value);
  const pageSize = items.length > 0 ? Math.min(items.length, DEFAULT_PAGE_SIZE) : DEFAULT_PAGE_SIZE;

  return {
    items,
    total: items.length,
    page: 1,
    page_size: pageSize,
    pages: 1,
    total_records: items.length,
    total_pages: 1,
    has_next: false,
    has_previous: false,
  };
}
